[do-not-merge] Bug/allow returning an error in dh

cammellos opened the pull request:
This commit changes the interface of `Encrypt` & `DH` to allow returning an error. It also remove the encrypted header version as wasn't maintained anymore and currently not used.

adambabik commented on crypto.go:
```diff
@@ -27,8 +27,8 @@ type DHPair interface {
 
-// Key is any 32-byte key. It's created for the possibility of pretty hex output.
-type Key [32]byte
+// Key is any 33-byte key. It's created for the possibility of pretty hex output.
```
Why do we extended it by 1 byte?

cammellos replied:
sorry, that's a left over, we change it to a slice, this implementation uses 32 bytes keys, while we need 33 bytes

adambabik commented on default_crypto.go:
```diff
@@ -32,48 +32,58 @@ func (c DefaultCrypto) GenerateDH() (DHPair, error) {
+		privKey [32]byte
+		pubKey  [32]byte
 	)
+	copy(privKey[:], dhPair.PrivateKey()[:32])
```
Not sure what's more dangerous: (1) copying less than 32 bytes or (2) panic. In this case, `[:32]` does not change anything because `privKey` is 32-bytes long anyway so no more data will be copied there. However, `PrivateKey` returns something shorter, this will panic.

https://play.golang.org/p/XC99saZV8Wa

cammellos replied:
Not sure how much we care about this implementation, is currently not used (we provide our own), so I'll be leaning toward just having it as an example

adambabik replied:
Ok, let's keep it then. At some point, we can refactor this to not have arrays at all if `DHPair` and `Key` do not require it.

smoelius replied:
Since you can now return an error, would it make sense to check that the length of `dhPair.PrivateKey()` is 32, and return an error if not?

adambabik commented on default_crypto.go:
```diff
@@ -32,48 +32,58 @@ func (c DefaultCrypto) GenerateDH() (DHPair, error) {
 	// The only error here is an entropy limit which won't be reached for such a short buffer.
 	_, _ = io.ReadFull(r, buf)
 
+	var rootKey Key = make(Key, 32)
```
Nitpick:
```suggestion
	rootKey := make(Key, 32)
```

adambabik commented on default_crypto.go:
```diff
@@ -101,7 +111,7 @@ func (c DefaultCrypto) Encrypt(mk Key, plaintext, ad []byte) []byte {
 	stream.XORKeyStream(ciphertext[aes.BlockSize:], plaintext)
 
-	return append(ciphertext, c.computeSignature(authKey[:], ciphertext, ad)...)
+	return append(ciphertext, c.computeSignature(authKey[:], ciphertext, ad)...), nil
```
This function cannot return any `error`. Is its signature dictated by some interface?

cammellos replied:
yes, this implementation is not used by us, we re-implement the methods, and in our case it might return an error

adambabik commented on keys_storage.go:
```diff
@@ -26,20 +27,21 @@ type KeysStorage interface {
 
 // Get returns a message key by the given key and message number.
 func (s *KeysStorageInMemory) Get(pubKey Key, msgNum uint) (Key, bool, error) {
+	index := fmt.Sprintf("%x", pubKey)
```
The same can be done with more explicit:
```suggestion
	index := hex.EncodeToString(pubKey)
```
This should be faster and have smaller memory footprint (probably does not matter much in this case but why not).

adambabik commented on keys_storage.go:
```diff
@@ -26,20 +27,21 @@ type KeysStorage interface {
 
 	// All returns all the keys
-	All() (map[Key]map[uint]Key, error)
+	All() (map[string]map[uint]Key, error)
```
Why do we need to change this?

cammellos replied:
we use slices now and can't be used as indexes for maps, it s only used for testing

cammellos commented:
@adambabik ready to be reviewed again!

adambabik commented on default_crypto.go:
```diff
@@ -32,48 +32,58 @@ func (c DefaultCrypto) GenerateDH() (DHPair, error) {
 	_, _ = io.ReadFull(r, buf)
 
+	rootKey := make(Key, 32)
+	var headerKey Key = make(Key, 32)
```
I added a comment to one line but here and in the line below we can simplify this like in the line 66.

adambabik commented on default_crypto.go:
```diff
@@ -142,10 +152,15 @@ func (c DefaultCrypto) deriveEncKeys(mk Key) (encKey Key, authKey Key, iv [16]by
 	// The only error here is an entropy limit which won't be reached for such a short buffer.
 	_, _ = io.ReadFull(r, buf)
 
+	var encKey Key = make(Key, 32)
```
The same here.

adambabik reviewed (approved):
A few comments to keep consistency. LGTM 👍

smoelius commented on default_crypto.go:
```diff
@@ -32,48 +32,58 @@ func (c DefaultCrypto) GenerateDH() (DHPair, error) {
+		pubKey  [32]byte
 	)
+	copy(privKey[:], dhPair.PrivateKey()[:32])
+	copy(pubKey[:], dhPub[:32])
```
Shouldn't that 32 be a 33?

Ref: https://crypto.stackexchange.com/a/8916

cammellos replied:
In this case it would be `32 bytes`, as this implementation uses `curve25519`, which compressed keys are only `32 bytes` https://tools.ietf.org/id/draft-josefsson-tls-curve25519-02.html#rfc.section.2.3 and  http://cr.yp.to/ecdh/curve25519-20060209.pdf . 

This caused a bit of a problem as we use instead `secp256k1` which compressed key has a length of `33 bytes`, but the type exposed was `32 bytes` (this implementation is not used at all in our production code, it's completely replaced by https://github.com/status-im/status-protocol-go/blob/master/crypto/ethereum_crypto.go ).

I have changed it to use a slice, which makes certain things easier and simplify the implementation, as it is converted to a slice when sending it over the wire because protobuf does not support fixed-size arrays.

But I could have also changed it to be `33 bytes`, which is the standard for elliptic keys, and just padded the array in this implementation (similarly as described in the rfc above).
Not sure what's best, what would you recommend?

smoelius replied:
Ah.  OK.

> Not sure what's best, what would you recommend?

I would use a slice, as you have.  I would be concerned that the padding approach would hide errors, e.g., an out-of-bounds read doesn't panic because it reads into the padding.

Thanks for the explanation.

cammellos replied:
thank you!

smoelius replied:
One more thought: would it make sense to check that the length of `dhPub` is 32 and return an error if not (and similarly for `dhPair.PrivateKey()`)?

I realize that this implementation is just an example. But I think, in the long term, there would be real benefit to verifying that all of `status-protocol-go`'s tests pass using both `EthereumCrypto` and `DefaultCrypto`.

cammellos replied:
sounds good, will do that, thanks

cammellos committed:
Remove encryted header version

cammellos committed:
Return error from DH & Encrypt

cammellos committed:
Use slices instead of fixed array for keys

In order to support multiple types of curves, this commits changes the
type of `Key` from `[32]byte` to `[]byte`. This is useful as most of the
eliptic keys have a size of a compressed key of `33 bytes` instead of
`32 bytes`, which the default implemetation uses (curve25519).

cammellos commented:
@smoelius Done, I have split the commits in 3 different (remove header encryption, return an error and change the type from `[32]byte` to `[]byte`).
I have also included your last suggestion of checking the length of the keys.
Thanks!
